client.go

```diff
@@ -270,13 +291,17 @@ func (c *Client) run() {
 	for {
 		select {
 		case call := <-calls:
-			if err := c.send(streamID, messageTypeRequest, call.req); err != nil {
-				call.errs <- err
-				continue
-			}
-
+			go func(streamID uint32, call *callRequest) {
+				requests <- streamCall{
```

Comment:
Why not just call `c.send` here directly, rather than pop over to another goroutine?

Reply:
That could result in multiple of this goroutines calling `c.send` concurrently, couldn't it?

Reply:
If you do keep this model, do you need to select here on `ctx.Done()` so that this goroutine doesn't hang forever?

(Alternatively maybe the other goroutine shouldn't select on `ctx.Done()` and should use some other scheme to determine when it's done.

Reply:
Good point, we don't have any way for these to be cleaned up if the connection closes.

Reply:
I think we need to keep a single sender goroutine that receives messages via channel and calls `c.send`. That will ensure we don't interleave the bits from multiple messages on the wire.

`ctx.Done()` seems to be the client's equivalent of the `done` channel on the server side, so I think that's probably most appropriate to select on to see when we should terminate.

Reply:
Answering your question, yes, that's true. I thought that was already possible but I was wrong. So that's out, I suppose.

A possible problem with this approach is that you've eliminated the backpressure on `calls`--if the socket is busy, we still keep processing messages from `calls`, allocating more goroutines without bound. Before, we would stop pulling messages off `calls`, which would allow someone select sending on `calls` (doubt this happens, though, didn't look yet). Also storing messages on `calls` is probably more memory and CPU efficient than storing them in blocked goroutines.

Not sure if that's a practical consideration.

Reply:
Ah, indeed we do select on sending to `calls`. So I think this is a problem worth solving.

I'd suggest trying to process `calls` directly in the new goroutine. You'll need to come up with a new scheme for synchronizing `waiters`--although you could play more games with channels, perhaps it's reasonable to just use a mutex in this case.

Reply:
I changed the send to a select with a `<-c.ctx.Done()` case, so we at least won't leak goroutines. I'll look at refactoring the rest of the flow to add back-pressure soon.

Reply:
I would tend to agree with @jstarks suggestion here, re: process calls in a new goroutine and use a mutex to sync waiters.